Port :erlang.binary_to_term/1 to JS

0bvim opened the pull request:
Closes #457

Implements `:erlang.binary_to_term/1` to decode Erlang External Term Format (ETF) binaries, enabling support for `Code.fetch_docs/1` in the browser.

### Implementation

Created a dedicated `EtfDecoder` class that handles ETF parsing with support for the following tag types:

**Integers:**
- SMALL_INTEGER_EXT (97), INTEGER_EXT (98)
- SMALL_BIG_EXT (110), LARGE_BIG_EXT (111)

**Atoms:**
- ATOM_EXT (100), SMALL_ATOM_EXT (115)
- ATOM_UTF8_EXT (118), SMALL_ATOM_UTF8_EXT (119)

**Collections:**
- BINARY_EXT (109) - binaries/strings
- SMALL_TUPLE_EXT (104), LARGE_TUPLE_EXT (105)
- NIL_EXT (106), STRING_EXT (107), LIST_EXT (108)
- MAP_EXT (116)

The decoder uses a recursive approach with offset tracking, leveraging `DataView` for efficient multi-byte value reading. Error handling matches Erlang's exact error messages.

### Testing

- **JavaScript unit tests:** 23 tests covering all implemented ETF types, edge cases, and error conditions
- **Elixir consistency tests:** 26 tests verifying behavior matches OTP exactly
- All existing tests (544) continue to pass

`describetag :binary` added to `binary_to_term` test. To allow the same use of `moduletag :consistency`. But focused to execute only specific kinds of test instead all tests in module.
Later we can improve test sections to gather all tests from the same type. 
If you think that it's a good idea, later I can add it to each section in a separated `pr`.

I guess that is a good idea add a `Makefile` to this repository and add targets to ease tests and etc... what you think about?

<img width="1017" height="154" alt="image" src="https://github.com/user-attachments/assets/0ddec292-2ced-4291-baec-e55fa4bd59d6" />

<img width="505" height="116" alt="image" src="https://github.com/user-attachments/assets/a02b24a0-cda3-493f-8985-cb9af57bc6ec" />

<img width="809" height="606" alt="image" src="https://github.com/user-attachments/assets/c635cf2a-3ea0-4879-884f-6f9ea36c1499" />


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Public Erlang binary decoder supporting full ETF term set, compressed input, and stricter validation/error messages.

* **Tests**
  * Large JS and Elixir test suites covering decoding, round-trip cases, compressed data, edge cases, and many error scenarios.

* **Tests (Helpers)**
  * Added an async test helper to assert expected boxed errors and centralized error-validation for test assertions.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

0bvim committed:
feat: add binary_to_term/1 implementation and EtfDecoder class

- Implemented the binary_to_term/1 function in erlang.mjs to decode binary representations of Erlang terms.
- Added a new EtfDecoder class in etf_decoder.mjs to handle the decoding of the Erlang External Term Format (ETF).
- Created comprehensive tests for binary_to_term/1 in both JavaScript and Elixir to ensure correct decoding of various Erlang data types, including integers, atoms, binaries, tuples, lists, and maps.
- Included error handling for invalid inputs and edge cases in the tests.

0bvim committed:
Enhance Erlang tests with describetags for binary functions

- Added @describetag :binary to the test cases for binary_to_list/1 and binary_to_term/1 in erlang_test.exs to improve test organization and clarity.

0bvim committed:
chore: removed describe tag, out of scope of this pr

bartblast commented on test/elixir/hologram/ex_js_consistency/erlang/erlang_test.exs:
```diff
@@ -1724,6 +1724,154 @@ defmodule Hologram.ExJsConsistency.Erlang.ErlangTest do
   end
 
+  describe "binary_to_term/1" do
+    @describetag :binary
```
We don't need this.

bartblast commented on test/javascript/erlang/etf_decoder_test.mjs:
```diff
@@ -0,0 +1,283 @@
+
+defineGlobalErlangAndElixirModules();
+
+describe("binary_to_term/1", () => {
```
These tests should go to `test/javascript/erlang/erlang_test.mjs `

bartblast commented on assets/js/erlang/etf_decoder.mjs:
```diff
@@ -0,0 +1,321 @@
+import Interpreter from "../interpreter.mjs";
+import Type from "../type.mjs";
+
+// Erlang External Term Format (ETF) Decoder
```
According to the https://www.erlang.org/doc/apps/erts/erl_ext_dist.html, the current implementation supports 15 out of ~30 ETF tags. The implementation is missing a few important term types that could appear in serialized Elixir/Erlang data:
- NEW_FLOAT_EXT (70) - Modern float format (8-byte IEEE floats)
- BIT_BINARY_EXT (77) - Non-byte-aligned bitstrings (e.g., <<1::3>>)
- EXPORT_EXT (113) - Captured function references (e.g., &Module.function/1)

We can skip deprecated/old formats and those related to anonymous functions, PIDs, and ports for now.

bartblast reviewed (changes requested):
Thank you for the PR! Added a few comments.

0bvim commented:
> Could you pull the new changes from the dev branch? (the CI reports the PR branch has conflicts)

ofc!

0bvim commented:
> > I guess that is a good idea add a `Makefile` to this repository and add targets to ease tests and etc... what you think about?
> 
> There are already a few mix tasks and test tags with similar function. To be honest I don't see a point in such a granular control of a specific subset of sets - you can achieve the same goal with running a test with specific line location in mix test or apply .only to the given describe block in JS tests.
> 
> Thanks for the suggestion! There are already a few mix tasks and test tags that provide similar functionality. To be honest, I don't see a strong need for such granular control of specific subsets of tests right now - you can achieve the same goal by:
> 
> * Running tests at a specific line location: `mix test path/to/test.exs:123`
> * Using the @moduletag or @tag filters in Elixir tests
> * Applying `.only` to a specific describe block in JS tests
> 
> If you find specific workflows that aren't well-supported by the existing tooling, we could revisit this. For now, let's keep the build tooling simple.

Nice!!! I'm used too use Makefile since my first project... So I ever want to use it... So your arguments are pretty nice. I'll get used to elixir tooling.

I will fix all requests and pull new changes from dev branch and update this `pr`!

bartblast commented:
Hey @0bvim, just checking in - do you need any help getting this PR across the finish line?

0bvim commented:
> Hey @0bvim, just checking in - do you need any help getting this PR across the finish line?

Hello @bartblast! 
Probably I'll do.
Sorry for delay in responding... this week was an chaos to me... :(
Today, sunday and next week I'll work on it... 
Now I have enough time... get it? :3

0bvim commented:
> @0bvim Do you think you’ll finish it by next Friday? I’m planning a big release with all the ported functions around January 25th.
@bartblast 
Ofc! I will ask your help if have any question!

0bvim committed:
wip: migrating etf decoder logic to erlang.mjs [skip ci]

0bvim committed:
chore: added main logic

- removed old file

0bvim committed:
fix: adjust mix aliases

- remove unused import from erlan.mjs
- adjust type hint in test function and renderer_dom
- 'mix f' formatted renderer_test.exs

0bvim committed:
chore: moved etf decoder mjs tests to erlang_test.exs

- removed describe tag from erlang_test.exs as recommended

0bvim committed:
chore: add limit check for decode atom, small atom, binary and string.

- fix typo in mix.exs js format.

0bvim committed:
fix: small atom bounds and add tests for extended ERL terms

Commit:
Add support for new ETF term tags

- Implement decoding for NEW_FLOAT_EXT, FLOAT_EXT, BIT_BINARY_EXT,
  NEW_REFERENCE_EXT, NEWER_REFERENCE_EXT, PID_EXT, NEW_PID_EXT,
  PORT_EXT, NEW_PORT_EXT and V4_PORT_EXT; add corresponding constants.
- Extend tests in Elixir and JS to cover these formats,
  including references, pids, ports, and bitstrings.

0bvim committed:
Update renderer.ex

0bvim committed:
Update test for LARGE_BIG_EXT encoding

0bvim committed:
Add tests for ETF tags and malformed binaries

0bvim committed:
feat: add EXPORT_EXT decoding for function captures

0bvim committed:
fix: decodeBinary condition

0bvim committed:
revert: revert renderer_Test.exs

0bvim committed:
fix: adjust code to be idiomatic and avoid credo warnings

bartblast reviewed (changes requested):
Thank you! :) Added some comments.

0bvim committed:
chore: add compressed term decoder

0bvim committed:
chore: adjust large big ext test to use a value that generate a
LARGE_BIG_EXT term.

0bvim committed:
chore: added tests to FLOAT_EXT

0bvim committed:
chore: add peer flags

0bvim commented:
> Hey @0bvim, how's it going with this PR? Any update on the progress? Need any help?

Hey @bartblast I guess that I'll do all the fix today. 
All progress that I've made is up to date with this branch. 
This week was very busy.

bartblast commented:
Hmm... that's odd - CI still shows formatting problems.

0bvim commented:
> Hmm... that's odd - CI still shows formatting problems.

@bartblast let me fix all the issues, revert needed content and format again and we see if it still happening.

0bvim committed:
wip: add compressed term and updated needed tests and functions

bartblast commented on assets/js/erlang/erlang.mjs:
```diff
@@ -572,6 +557,646 @@ const Erlang = {
+      );
+    }
+
+    // ETF tag constants
```
Thanks @0bvim for tackling this, I know async refactors can be a pain! 🙏
One request: could you move all the decompression-related code from the Utils class directly into the ported function itself? The reason is that we're aggressively optimizing bundle size through call graph analysis - if it lives in Utils, it'll get included in every Hologram application even if they never use this function. By keeping it inside the ported function, the Hologram compiler can tree-shake it out when it's not needed.

0bvim committed:
chore: adjusted tests to handle async.

- Create async assertBoxedErrorAsync function.

0bvim committed:
chore: Move zlibInflate into Erlang binary_to_term

Remove Utils.zlibInflate and add equivalent function inside the
binary_to_term implementation. Update callsites and add a test to
validate internal zlibInflate behavior for compressed terms.

0bvim committed:
revert: renderer_test.exs

0bvim committed:
Use preferred_cli_env in mix.exs

Replace cli with preferred_cli_env in the project config.
Update format aliases: use 'cmd cd test/features && mix format' and
broaden the prettier scripts glob.

0bvim committed:
Validate bit count for bit binary decoding

Raise ArgumentError when BIT_BINARY_EXT 'bits' byte is not 1..8. Update
tests to use bits=8 for the full-byte case and apply minor formatting
fixes in test fixtures.

bartblast commented:
@0bvim I'd like to keep the PR focused on the ported function only.

bartblast commented on assets/js/erlang/erlang.mjs:
```diff
@@ -599,6 +599,782 @@ const Erlang = {
+
+      if (offset + 5 + length > bytes.length) {
+        Interpreter.raiseArgumentError(
+          `bit binary length exceeds available bytes: ${length}`,
```
Same issue here - custom error message doesn't match Erlang's output.

bartblast commented on assets/js/erlang/erlang.mjs:
```diff
@@ -599,6 +599,782 @@ const Erlang = {
+    const decodeFloatExt = (dataView, bytes, offset) => {
+      // FLOAT_EXT: 31-byte null-terminated string (deprecated format)
+      if (offset + 31 > bytes.length) {
+        Interpreter.raiseArgumentError("float string exceeds available bytes");
```
Same issue - custom error message doesn't match Erlang's output.

bartblast commented on test/elixir/hologram/ex_js_consistency/erlang/erlang_test.exs:
```diff
@@ -1946,6 +1946,379 @@ defmodule Hologram.ExJsConsistency.Erlang.ErlangTest do
+                   {:erlang, :binary_to_term, [:test]}
+    end
+
+    test "raises ArgumentError if argument is a non-binary bitstring" do
```
There's no corresponding JS test, please add:
```javascript
it("raises ArgumentError if argument is a non-binary bitstring", async () => {
  const bits = Bitstring.fromBits([1, 0, 1]);

  await assertBoxedErrorAsync(
    () => binary_to_term(bits),
    "ArgumentError",
    Interpreter.buildArgumentErrorMsg(1, "not a binary")
  );
});
```

bartblast commented on test/javascript/erlang/erlang_test.mjs:
```diff
@@ -2753,6 +2754,1053 @@ describe("Erlang", () => {
+        // Values 1 to 300, but modulo 256 for values > 255
+        const elementBytes = [];
+        for (let i = 1; i <= arity; i++) {
+          elementBytes.push(97, i & 0xff); // SMALL_INTEGER_EXT + value (wraps at 256)
```
The test creates element values with `i & 0xff`, which means values wrap at 256 (e.g., element 256 has value 0, element 257 has value 1). This is fine for testing the decoder, but a comment explaining this would be helpful for future maintainers.

**Action:** Add a comment:
```javascript
// Values 1 to 300, but SMALL_INTEGER_EXT only supports 0-255,
// so we use modulo 256 to keep all elements as SMALL_INTEGER_EXT
elementBytes.push(97, i & 0xff);
```

bartblast commented on test/javascript/support/helpers.mjs:
```diff
@@ -93,6 +93,57 @@ export function assertBoxedError(
   }
 }
 
+export async function assertBoxedErrorAsync(
```
The new `assertBoxedErrorAsync` function duplicates ~50 lines of validation logic from the existing `assertBoxedError`. The only difference is `await asyncCallable()` vs `callable()`.

Please extract the common validation logic into a shared helper, e.g.:

```javascript
function validateBoxedError(error, expectedErrorType, expectedErrorMessage) {
  const isRegex = expectedErrorMessage instanceof RegExp;
  const expectedMessageDisplay = isRegex
    ? expectedErrorMessage.toString()
    : expectedErrorMessage;

  const failMessagePrefix = `\nexpected:\n${expectedErrorType}: ${expectedMessageDisplay}\n`;

  if (!error) {
    assert.fail(failMessagePrefix + "but got no error");
  }

  if (!(error instanceof HologramBoxedError)) {
    assert.fail(
      failMessagePrefix + `but got:\n${error.name}: ${error.message}`,
    );
  }

  const receivedErrorType = Interpreter.getErrorType(error);
  const receivedErrorMessage = Interpreter.getErrorMessage(error);

  const typeMatches = receivedErrorType === expectedErrorType;
  const messageMatches = isRegex
    ? expectedErrorMessage.test(receivedErrorMessage)
    : Interpreter.isStrictlyEqual(
        error.struct,
        Type.errorStruct(expectedErrorType, expectedErrorMessage),
      );

  if (!typeMatches || !messageMatches) {
    assert.fail(
      failMessagePrefix +
        `but got:\n${receivedErrorType}: ${receivedErrorMessage}`,
    );
  }
}

export function assertBoxedError(callable, expectedErrorType, expectedErrorMessage) {
  let error;
  try { callable(); } catch (e) { error = e; }
  validateBoxedError(error, expectedErrorType, expectedErrorMessage);
}

export async function assertBoxedErrorAsync(asyncCallable, expectedErrorType, expectedErrorMessage) {
  let error;
  try { await asyncCallable(); } catch (e) { error = e; }
  validateBoxedError(error, expectedErrorType, expectedErrorMessage);
}
```

bartblast reviewed (changes requested):
Thanks for the work on this - it's a substantial implementation that covers a lot of ground.

I've left some comments on specific issues. The main areas are:
- Inconsistent error messages (some use custom messages instead of `buildArgumentErrorMsg`)
- `console.error` left in production code
- Missing REFERENCE_EXT support
- Some test coverage gaps

I didn't cover everything in detail - please make sure the behaviour is fully consistent between the JS and Elixir implementations. The best way to achieve this is to ensure the JS tests and Elixir consistency tests are **exactly mirrored**: same test names, same order, same arguments. This makes it much easier to verify correctness and maintain going forward.

0bvim committed:
feat: handle binary_to_term

0bvim committed:
Use Interpreter.buildArgumentErrorMsg in tests

Replace hard-coded argument-error strings with
Interpreter.buildArgumentErrorMsg
calls in erlang_test.mjs. Also normalize spacing in imports and
object-literal
entries for consistent formatting.

0bvim committed:
chore: Add Erlang external-term format consistency tests

0bvim committed:
chore: add bitstring js test

0bvim committed:
chore: fix error message to match with erlang error message.

0bvim committed:
refactor: Extract validateBoxedError helper

bartblast commented:
Thank you @0bvim! 💜 🎉

0bvim commented:
> Thank you @0bvim! 💜 🎉

You're welcome! 
I've got a new job now and I'm busy learning new stuff and so on. But I want to contribute more later. 
Congrats for sponsor and partnership with curiosum, well deserved man!
